src/v/storage/tiered_storage.cc

```diff
@@ -0,0 +1,516 @@
+    // If the manifest name in S3 matches the old manifest name it will
+    // be overwriten and some data may be lost as a result.
+    vassert(
+      upl_result == cloud_storage::upload_result::success,
+      "Can't upload new manifest {} after recovery",
+      target.get_manifest_path());
+
+    // TODO (evgeny): clean up old manifests on success. Take into account
+    //                that old and new manifest names could match.
+
+    // Upload topic manifest for re-created topic (here we don't prevent
+    // other partitions of the same topic to read old topic manifest if the
+    // revision is different).
+    if (mat.topic_manifest.get_revision() != _ntpc.get_revision()) {
+        mat.topic_manifest.set_revision(_ntpc.get_revision());
+        upl_result = co_await _remote->upload_manifest(
+          _bucket, mat.topic_manifest, _rtcnode);
+        if (upl_result != cloud_storage::upload_result::success) {
+            // That's probably fine since the archival subsystem will
+            // re-upload topic manifest eventually.
+            vlog(
+              _ctxlog.warn,
+              "Failed to upload new topic manifest {} after recovery",
+              target.get_manifest_path());
+        }
+    }
+    co_return;
+}
+
+ss::future<> partition_downloader::download_log_with_capped_size(
+  const offset_map_t& offset_map,
+  const cloud_storage::manifest& manifest,
+  const std::filesystem::path& prefix,
+  size_t max_size) {
+    vlog(_ctxlog.info, "Starting log download with size limit at {}", max_size);
+    gate_guard guard(_gate);
+    size_t total_size = 0;
+    for (auto it = offset_map.rbegin(); it != offset_map.rend(); it++) {
+        const auto& meta = it->second.meta;
+        if (total_size > max_size) {
+            vlog(
+              _ctxlog.debug,
+              "Max size {} reached, skipping {}",
+              total_size,
+              it->second.full_path);
+            break;
+        } else {
+            vlog(
+              _ctxlog.debug,
+              "Downloading {}, total log size {}",
+              it->second.full_path,
+              total_size);
+        }
+        auto fname = cloud_storage::segment_name(it->second.full_path);
+        co_await download_file(fname, manifest, prefix);
+        total_size += meta.size_bytes;
+    }
+}
+
+ss::future<> partition_downloader::download_log_with_capped_time(
+  const offset_map_t& offset_map,
+  const cloud_storage::manifest& manifest,
+  const std::filesystem::path& prefix,
+  model::timestamp_clock::duration retention_time) {
+    vlog(
+      _ctxlog.info,
+      "Starting log download with time limit of {}s",
+      retention_time.count() / 1000);
+    gate_guard guard(_gate);
+    auto time_threshold = model::to_timestamp(
+      model::timestamp_clock::now() - retention_time);
+    for (auto it = offset_map.rbegin(); it != offset_map.rend(); it++) {
+        const auto& meta = it->second.meta;
+        if (
```

Comment:
In this and the _capped_size function, the calls to `download_file` are being run in serial -- looks like it should be possible to compose the list of desired offsets first, and then dispatch the fetches in parallel.

Reply:
The reasoning behind this is that during disaster recovery we will recover many partitions at the same time on every node. So there is no need to make individual partition recovery parallel.
On the other hand, when we're running individual topic recovery parallelization might be useful. But also, both recovery and upload share the same connection pool. The pool is fixed size and, for instance, might have 10 connection per shard. If we start downloading a lot of segments in parallel during topic recovery we might starve upload. 
I don't have a strong opinion here though. Upload might wait for a while, no big deal.